test/blackbox-tests/test-cases/optional-executable.t/run.t

```diff
@@ -52,3 +52,110 @@ The following command should fail because the executable is not optional:
                   ^^^^^^^^^^^^^^
   Error: Library "does-not-exist" not found.
   [1]
+
+A strange behavior discovered in #4786. Dune would ignore an executable if any
+of its dependencies were optional.
+
+  $ mkdir optional-binary
+  $ cd optional-binary
+  $ cat >dune-project <<EOF
+  > (lang dune 3.0)
+  > (package (name myfoo))
+  > EOF
+
+  $ mkdir exe
+  $ cat >exe/bar.ml <<EOF
+  > print_endline "hello world"
+  > EOF
+  $ cat >exe/dune <<EOF
+  > (executable (public_name dunetestbar) (name bar) (libraries foo))
+  > EOF
+
+  $ mkdir lib
+  $ cat >lib/dune <<EOF
+  > (library (name foo) (libraries xxx-does-not-exist) (optional) (modules ()))
+  > EOF
+
+  $ cat >dune <<EOF
+  > (rule
+  >  (alias run-x)
+  >  (action (echo %{exe:bar.exe})))
+  > EOF
+
+  $ dune build @run-x
+  Error: No rule found for bar.exe
```

Comment:
"No rule found"? That's still a pretty broken behavior, yeah? (which I agree is less broken than before this PR)
I think it's broken because the error message is bad (obviously), but more subtly because you can probably define a custom rule that generates bar.exe and then dune probably succeeds despite the dune files being crazy.

Reply:
> "No rule found"? That's still a pretty broken behavior, yeah? (which I agree is less broken than before this PR)
I think it's broken because the error message is bad (obviously)

I'll see if I can improve it, but it does not seem so easy at first glance.

>  but more subtly because you can probably define a custom rule that generates bar.exe and then dune probably succeeds despite the dune files being crazy.

Could you elaborate on this more?  I custom rule wouldn't be enough, you would need to include the file in `PATH` via public_name or the `env` stanza. In which case the desired executable will be found - as expected.

Reply:
> I [think] custom rule wouldn't be enough, you would need to include the file in PATH via public_name or the env stanza. 

Well, maybe, but at the very least the error message will be different in a confusing way, and also the following will work:

```
(rule (target foo.exe) ...)
(rule (target another-file.txt) (deps foo.exe) ...)
```

Anyway, that's not really a problem by itself, it just highlighting that there's some shaky ground here.